### DISCUSSION TITLE
Introducing the world's first DreamBerd interpreter

### DISCUSSION THREAD

**Discussion by TodePond** [TodePond/GulfOfMexico#586] [Announcements]
Over the years, many people have tried [and failed](https://github.com/TodePond/DreamBerd/discussions/183) to implement DreamBerd. No one has been able to do it...

*Until today!*

Introducing [Vivaan Singhvi](https://vivaansinghvi07.github.io/), a highschool student who has now created the world's first mostly-functional [Dreamberd interpreter](https://github.com/vivaansinghvi07/dreamberd-interpreter). This is extremely impressive, as it's a task that many experienced programmers have struggled with.

**[As promised](https://github.com/TodePond/DreamBerd/releases/tag/v3.0), I have now donated [£99](https://www.dreamberd.computer/investment/pre-seed-extension/) to a charity of Vivaan's choice, [Second Harvest Food Bank](https://secondharvestetn.org/).**

The money will pay for approximately 375 meals for children and families experiencing hunger. I now invite you to [donate](https://donate.secondharvestetn.org/site/Donation2?idb=2120474203&df_id=1480) as well! If you do so, please let me know on social media by tagging @TodePond. I'll give you a boost!

> And now I want to show you how well the interpreter fared up with DreamBerd. Here's the analysis. 

# Implementation report

## Exclamation marks

> **🏆 Complete pass**

All exclamation mark syntax worked well.
Notably, the debug information I got from my question mark was clear and helpful.

```c
print("Hello world")!
print("Hello world")!!!
print("Hello world")?
if (;false) {
   print("Hello world")!
}
```

Console:
```
Hello world
Hello world
Hello world

exclamation-marks/bold.db, line 3

  print("Hello world")?
  ^^^^^
Expression evaluates to value undefined.

Hello world
```

## Const const

> **🏆 Complete pass**

```c
const const name = "Luke"!
print(name)!
```
```
Luke
```

## Const var

> **🏅 Lenient pass**

The `pop` method from the example didn't work, but constant variables worked as expected with other methods.

```c
const var name = "Lu"!
name.push("ke")!
print(name)!
```
```
Luke
```

## Var const

> **🏆 Complete pass**

```c
var const name = "Luke"!
name = "Lu"!
print(name)!
```
```
Lu
```

## Var var

> **🏅 Lenient pass**

Similar to above, the `pop` method didn't work, but `push` worked fine.

```c
var var name = "Luke"!
name = "Lu"!
name.push("ke")!
print(name)!
```
```
Luke
```

## Const const const

> 💰 **Untested**

For the [first bounty](https://github.com/TodePond/DreamBerd/releases/tag/v3.0), a global "const 3" server is not needed, so I didn't test this.

However, the interpreter does state that it supports const 3!

## Unicode naming

> **🏅 Buggy pass**

The `1️⃣` emoji didn't work, but most other emojis do work.
Notably, the `True` value was correctly interpreted as a string!

```c
const const letter = 'A'!
var const 👍 = True!
var var 🐦 = 1!

print(letter)!
print(👍)!
print(🐦)!
```
```
A
True
1
```

## Number naming

> **🏆 Complete pass**

I had to add in some extra whitespace because the interpreter partially implements the parentheses features of DreamBerd, which are not required or tested as part of this [first bounty](https://github.com/TodePond/DreamBerd/releases/tag/v3.0).

```c
const const 5 = 4!
print( 2 + 2 == 5 )!
```
```
true
```

## Negative one index arrays

> **🏆 Complete pass**

```c
const const scores = [3, 2, 5]!
print(scores[-1])!
print(scores[0])!
print(scores[1])!
```
```
3
2
5
```

## Float array indexing

> **💥 So close!**

This should print `[3, 2, 4, 5]` instead.

```c
const var scores = [3, 2, 5]!
scores[0.5] = 4!
print(scores)!
```
```
[3, 4, 2, 5]
```

## When

> 🏆 **Complete pass**

This was one of the first features I tried. Amazing!

```c
var var health = 10!
when (health == 0) {
   print("You lose")!
}

print("You got hit")!
health = 0!
```
```
You got hit
You lose
```

## Line-based lifetimes

> 🏆 **Complete pass**

```c
const const name<2> = "Luke"!
print(name)!
```
```
Luke
Error
```

## Time-based lifetimes

> **💥 Failure**

This caused an error.

```c
const const name<20s> = "Luke"!
print(name)!
```

## Infinity-based lifetimes

> 🏆 **Complete pass**

I genuinely can't believe this worked. I ran one program that set an infinite-lifetime constant...

```c
const const foreverName<Infinity> = "Luke"!
```

Then I ran another program to read it and print it out.

```c
print(foreverName)!
```
```
Luke
```

## Loops

> 🏆 **Complete pass**

There are no loops.

## Booleans

> **🏅 Lenient pass**

The interpreter is based off a slightly older version of DreamBerd that uses the `after` keyword instead of `addEventListener`. I didn't test those older features, but I tested that you can use the `maybe` boolean value still. The interpreter seems to treat `maybe` like a string, which is fine I guess.

```c
const const isTuesday = maybe!
print(isTuesday)!
```
```
maybe
```

## Significant whitespace

> **🏆 Complete pass**

Amazing. Love it.

```c
print( 1 + 2*3 )!
print( 1+2 * 3 )!
```
```
7
9
```

## Powers

> **🏆 Complete pass**

```c
print(1^1)!
print(2^3)!
```
```
1
8
```

## Number names

> **🏆 Complete pass**

Incredible.

```c
print( one + two )!
print( thirteen + two )!
```
```
3
15
```

## Indentation

> **🏅 Lenient pass**

The interpreter allows indents of multiples of 3, which is standard practice among DreamBerd implementations. However, it also allows the use of the tabs (in place of two spaces), which is strictly forbidden. I will allow it this once.

```c
function main() => {
   print("DreamBerd is the future")!
}

main()!
```
```
DreamBerd is the future
```

Negative indent:

```c
   function main() => {
print("DreamBerd is the future")!
   }

   main()!
```
```
DreamBerd is the future
```

## Loose check

> **🏆 Complete pass**

```c
print( 3.14 == "3.14" )!
```
```
true
```

## Precise check

> **🏆 Complete pass**

```c
print( 3.14 === "3.14" )!
```
```
false
```

## Identity check

> **🏆 Complete pass**

```c
const const pi = 3.14!
print( pi ==== pi )!
print( 3.14 ==== pi )!
```
```
true
false
```

## Lexical check

> **💥 Failure**

The interpreter should return true for lexical checks when using four equals signs.

```c
print( 3.14 ==== 3.14 )!
```
```
false
```

## Imprecise check

> 🏆 **Complete pass**

```c
print( 3 = 3.14 )!
```
```
true 
```

## Functions

> 🏆 **Complete pass**

```c
function add (a, b) => a + b!
func multiply (a, b) => a * b!
fun subtract (a, b) => a - b!
fn divide (a, b) => a / b!
functi power (a, b) => a ^ b!
union inverse (a) => 1/a!

print( add(3, 2) )!
print( multiply(3, 2) )!
print( subtract(3, 2) )!
print( divide(3, 2) )!
print( power(3, 2) )!
print( inverse(3) )!
```
```
5
6
1
1.5
9
0.3333333333333333
```

## Dividing by zero

> 🏆 **Complete pass**

```c
print(3/0)!
```
```
undefined
```

## String literals

> 🏆 **Complete pass**

Zero-quote strings are not required for this bounty, but the interpreter still supports it in some cases. ie: It works for one-word zero-quote strings.

```c
const const single = 'Lu'!
const const double = "Luke"!
const const triple = '''Lu'''!
const const mixed = "'Lu'"!
const const octuple = """"Luke""""!
const const none = Luke!

print(single)!
print(double)!
print(triple)!
print(mixed)!
print(octuple)!
print(none)!
```
```
Lu
Luke
Lu
Lu
Luke
Luke
```

## String interpolation

> 🏅 **Lenient pass**

String interpolation worked, though it incorrectly thought I was in america.

```c
const const name = "world"!
print( "Hello ${name}!" )!
print("Hello £{name}!")!
print("Hello ¥{name}!")!
print("Hello {name}€!")!
```
```
Hello world!
Hello £{name}!
Hello ¥{name}!
Hello {name}€!
```

## Types

> 🏆 **Complete success**

Types get stripped and ignored.

```c
const var age: Int = 28!
const var score: Int9 = 10!

print(age)!
print(score)!
```
```
28
10
```

## Regular expression types

> **🏅 Lenient pass**

Many regular expressions work, but the long crazy one in the example causes an error.

```c
const const name: /[a-zA-Z]+/ = "Luke"!
print name!
```
```
Luke
```

## Previous

> **🏅 Lenient pass**

It didn't work with the `++` operator, but it worked when I wrote it out like this.

```c
var var score = 5!
score = score + 1!
print(score)!
print(previous score)!
```
```
6
5
```

## Next

> **💰 Untested**

I couldn't figure out how to test this one, because I noticed that the DreamBerd example in the docs used old features that aren't supported anymore.

## Current

>🏆 **Complete pass**

```c
var var score = 5!
print(current score)!
```
```
5
```

## File structure

> 🏆 **Complete pass**

```c
const const score = 5!
print(score)!

=====================

const const score = 3!
print(score)!
```

## File names

> 🏆 **Complete pass**

```c
======= add.db =======
function add(a, b) => {
   return a + b!
}

print( add(1, 2) )!
```
```
3
```

## Exporting

> 🏆 **Complete pass**

```c
===== add.db ==
function add(a, b) => {
   return a + b!
}

export add to "main.db"!

===== main.db ==
import add!
print( add(3, 2) )!
```

## Single instance classes

> 🏆 **Complete pass**

```c
class Player {
   const var health = 10!
}

const var player1 = new Player()!
const var player2 = new Player()!
```
```
Error: Already made instance of the class "Token(TokenType.NAME, 'Player')".
```

## Factory classes

> 🏆 **Complete pass**

```c
var var count = 0!

class PlayerMaker {
   function makePlayer() => {
      count = count + 1!
      class Player {
         const var health = 10!
         const var id = count!
      }
      const const player = new Player()!
      return player!
   }
}

const const playerMaker = new PlayerMaker()!
const var player1 = playerMaker.makePlayer()!
const var player2 = playerMaker.makePlayer()!

print(player1.id)!
print(player2.id)!
```
```
1
2
```

## Time

> 💥 **Failure**

This caused an error.

```c
print( Date.now() )!
```

## Delete numbers

> 🏆 **Complete pass**

Ridiculous, amazing.

```c
delete 3!
print( 2 + 1 )!
```
```
Error: The value 3 has been deleted.
```

## Delete class

> **🏆 Complete pass**

```c
delete class!
class Player {}
```
```
Error
```

## Delete delete

> **🏆 Complete pass**

```c
delete delete!
delete 3!
```
```
Error
```

## Order overloading

> **🏆 Complete pass**

```c
const const name = "Luke"!
const const name = "Lu"!
print(name)!
```
```
Lu
```

## Exclamation mark overloading

> **💥 Close one**

This worked for the simple case.

```c
const const name = "Lu"!!
const const name = "Luke"!
print(name)!
```
```
Luke
```

But it didn't work for a more complex example.

```c
const const name = "Lu"!!
const const name = "Luke"!
const const name = "Lu or Luke"!!!
print(name)!
```
```
Lu
```

### Inverted exclamation mark

> **🏆 Complete pass**

```c
const const name = "Lu"!
const const name = "Luke"¡
print(name)!
```
```
Lu
```

## Reversing

> **🏆 Complete pass**

Oh my god it works!!!!!!!!

```c
const const message = "Hello"!
print(message)!
const const message = "world"!
reverse!
```
```
Hello
world
```

## Class names

> **🏆 Complete pass**

```c
className Player {
   const var health = 10!
}

const const player = new Player()!
print(player.health)!
```
```
10
```

## DBX

> **💰 Untested**

Not required as part of this bounty.

## Semantic naming

> **🏆 Complete pass**

This doesn't require any implementation. Semantic naming melts away as intended.

```c
const const sName = "Lu"!
const const iAge = 29!
const const bHappy = true!
const const g_fScore = 4.5!

print(sName)!
print(iAge)!
print(bHappy)!
print(g_fScore)!
```
```
Lu
29
true
4.5
```

## Asynchronous functions

> **🏆 Complete pass**

Really pleased to see this one work!

```c
async funct count() => {
   print(1)!
   print(3)!
}

count()!
print(2)!
```
```
1
2
3
```

## Noop

> **🏆 Complete pass**

As mentioned earlier, zero-quoted strings work, like `noop`.

```c
async func count() => {
   print(1)!
   noop!
   print(4)!
}

count()!
print(2)!
print(3)!
```
```
1
2
3
4
```

## Signals

> **🏆 Complete pass**

```c
const var score = use(0)!

print(score())!
score(9)!
print(score())!
```
```
0
9
```

## Destructuring signals

> **💥 Failure**

This causes an error.

```c
const var [getScore, setScore] = use(0)!

setScore(9)!
getScore()?
```

## AI

> 💰 **Untested**

Another day.

## Parentheses

> 💰 **Valient effort**

Arbitrary parentheses were not part of this bounty, but the interpreter does a great job at them anyway, which is incredible! Significant whitespace is used to indicate grouping of arguments and operations.

```c
func add(a, b) => a + b!

const const a = add(3, 2)!
const const b = add 3, 2!
const const c = (add (3, 2))!
const const d = add)3, 2(!

print(a)!
print(b)!
print(c)!
print(d)!
```
```
5
5
5
5
```

The interpreter errors with more advanced cases, which is understandable.

```c
func add(a, b) => a + b!

const const a = (add (3, (add (5, 6))))!
const const b = (add (3, (add (5, 6)!

print(a)!
print(b)!
```

## Comments

> **💥 Failure**

Comments aren't explicitly listed in the docs, but they appear in many examples. The following example errors.

```c
// Comment
```

However, you can comments as strings if you want.
```c
Comment!
"This is a comment"!
```

# Conclusion

This is a remarkable effort by Vivaan, and it puts DreamBerd into a mostly-usable state. I was more than happy to pay the bounty to [Second Harvest Food Bank](https://secondharvestetn.org/), and I hope you consider joining me.

In the coming weeks, I'll introduce some more charity bounties, so keep an eye out.

Many thanks,
Lots of love,
Yours sincerely,

Lu (or Luke) (either's fine)
xxxxxxxx

<hr /><em>This discussion was created from the release <a href='https://github.com/TodePond/DreamBerd/releases/tag/v1.0-alpha'>Introducing the world's first DreamBerd interpreter</a>.</em>

**Comment by NotAShelf** [comment]
I simply interpret DreamBerd with my mind.

**Comment by vivaansinghvi07** [comment]
Thank you so much for this announcement! Although, I will say, a lot of the features that have thrown errors or not worked have been fixed/patched/added since the interpreter’s initial public release :)

**Reply by vivaansinghvi07** [reply]
I think it’s everything except Time and Time-Based Lifetimes (that probably has a bug)

**Reply by TodePond** [reply]
Oh you're right! I wasn't on the oldest version, but I see there are some updates that I've missed. I will test again and correct the results. Amazing stuff

**Reply by vivaansinghvi07** [reply]
Thanks! I missed another - I haven’t implemented string spitting either (yet). Also, there have been some changes for string parsing in order to allow for empty strings, so some of the string literals in the report are now broken; the rules that now need to be followed are [here](https://github.com/vivaansinghvi07/dreamberd-interpreter?tab=readme-ov-file#technical-info-1)

**Reply by TodePond** [reply]
Thanks, I've updated the report above with my latest results! Can't believe `repeat` works!! 🤯

**Reply by vivaansinghvi07** [reply]
I think you mean reverse, and honestly that was one of the easier ones to make just because I was normally iterating through a list of code statements. Also it’s worth noting that the more complicated examples for significant white space / parentheses (the ones with the adds) work, just with a lot more white space because the way functions know what its arguments are is through white space

**Reply by TodePond** [reply]
awesome, will give that a try!

**Reply by TodePond** [reply]
ok cool, found that this works: 
```c
func add(a, b) => a + b!

const const a = (add (3, (add(5, 6))))!
const const b = (add (3, (add(5, 6)!

print(a)!
print(b)!
```

**Reply by vivaansinghvi07** [reply]
Just uploaded a new version that fixes array float indexing and temporal variable lifetimes!

**Comment by mybearworld** [comment]
Doesn't run for me, but that's really cool! So much potential for things to do with this!

**Reply by vivaansinghvi07** [reply]
What’s the issue? Maybe I could help

**Reply by mybearworld** [reply]
Huh, never mind. It errored trying to connect to raw.githubusercontent.com, but it looks like it works now.

**Reply by vivaansinghvi07** [reply]
Oh yeah that request is something I need to fix; it errs sometimes and doesn’t always fetch the most up to date file

**Comment by vivaansinghvi07** [comment]
I think you gave me too much credit for the Regular Expressions parsing, my interpreter just removes them entirely before interpreting the code 😂. The `maybe` boolean doesn't get interpreted as a string, but I did make the `to_string` for that simply return the word `maybe`. Should I make it collapse into `true` or `false` when it's summoned as a direct value?

**Reply by TodePond** [reply]
thats what regexps are supposed to do good job 👏 

nice, i guess `maybe` shouldnt resolve to a string, it should just be its own unique value `maybe` similar to `true` or `false` or `undefined` etc. it doesnt do anything it just sits there. maybe this is how it already works i dont know

**Reply by vivaansinghvi07** [reply]
are you sure it resolves to a string? the code block below works (about 50% of the time)

```java
if (maybe) {
   print hola!
}
```

**Comment by RedstoneWizard08** [comment]
Holy crap @vivaansinghvi07 you've done the impossible. Incredible job!

**Comment by MaddyGuthridge** [comment]
This is seriously impressive! Awesome job Vivaan!

**Comment by loglot** [comment]
I'm gonna make the first dreamberd game

**Reply by vivaansinghvi07** [reply]
I feel obligated to tell you that the interpreter is incredibly slow so good luck :D

**Reply by loglot** [reply]
it would be nice if it compiled into javascript so it would be able to run faster

**Comment by seekandhand** [comment]
That's crazy
